sql/src/test/java/io/crate/planner/WhereClauseOptimizerTest.java

```diff
@@ -0,0 +1,161 @@
+import io.crate.metadata.RowGranularity;
+import io.crate.metadata.doc.DocTableInfo;
+import io.crate.test.integration.CrateDummyClusterServiceUnitTest;
+import io.crate.testing.SQLExecutor;
+import org.apache.lucene.util.BytesRef;
+import org.junit.Before;
+import org.junit.Test;
+
+import static io.crate.testing.SymbolMatchers.isLiteral;
+import static io.crate.testing.TestingHelpers.isDocKey;
+import static java.util.Collections.singletonList;
+import static org.hamcrest.Matchers.contains;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.empty;
+import static org.hamcrest.Matchers.is;
+
+public class WhereClauseOptimizerTest extends CrateDummyClusterServiceUnitTest{
+
+    private SQLExecutor e;
+
+    @Before
+    public void setUpExecutor() throws Exception {
+        e = SQLExecutor.builder(clusterService)
+            .addTable("create table bystring (name string primary key, score double) " +
+                      "clustered by (name) ")
+            .addTable("create table clustered_by_only (x int) clustered by (x)")
+            .addTable("create table tclustered (x int) clustered by (x)")
+            .addPartitionedTable(
+                "create table parted (id int, date timestamp) " +
+                "partitioned by (date)"
+            )
+            .addPartitionedTable(
+                "create table parted_pk (id int primary key, date timestamp primary key) " +
+                "partitioned by (date)",
+                new PartitionName("parted_pk", singletonList(new BytesRef("1395874800000"))).asIndexName(),
+                new PartitionName("parted_pk", singletonList(new BytesRef("1395961200000"))).asIndexName(),
+                new PartitionName("parted_pk", singletonList(null)).asIndexName()
+            )
+            .build();
+    }
+
+    private WhereClauseOptimizer.DetailedQuery optimize(String statement) {
+        QueriedTable queriedTable = e.analyze(statement);
+        EvaluatingNormalizer normalizer = new EvaluatingNormalizer(
+            e.functions(),
+            RowGranularity.CLUSTER,
+            null,
+            queriedTable.tableRelation()
+        );
+        return WhereClauseOptimizer.optimize(
+            normalizer,
+            queriedTable.where().queryOrFallback(),
+            ((DocTableInfo) queriedTable.tableRelation().tableInfo()),
+            e.getPlannerContext(clusterService.state()).transactionContext()
+        );
+    }
+
+    @Test
+    public void testFilterOn_IdDoesNotProduceDocKeysWithOnlyClusteredBy() {
+        WhereClauseOptimizer.DetailedQuery query = optimize(
+            "select * from tclustered where _id = '1'");
+        assertThat(query.docKeys().isPresent(), is(false));
+    }
+
+    @Test
+    public void testFilterOn_IdOnPartitionedTableDoesNotResultInDocKeys() {
+        WhereClauseOptimizer.DetailedQuery query = optimize(
+            "select * from parted where _id = '1'");
```

Comment:
can we use `id` and rename this test to `testFilterOnPrimaryKeyOnlyOnPartitionedTableDoesNotResultInDocKeys` so make it more clear? afaik usage of `_id` is a bit irrelevant here and may confuse...

Reply:
I had assumed that `_id` is special because it's an implicit primary key if the table has no other primary key definition. If I add an `id` column that is a primary key, the partitioned by column also has to become a primary key and then the test would effectively test that partial pk-column presence doesn't result in docKeys. That's already covered.

But again I'm not sure about the intention of the original tests.